compiler/src/dotty/tools/backend/jvm/BCodeHelpers.scala

```diff
@@ -574,43 +577,54 @@ trait BCodeHelpers extends BCodeIdiomatic with BytecodeWriters {
      */
     def addForwarders(jclass: asm.ClassVisitor, jclassName: String, moduleClass: Symbol): Unit = {
       assert(moduleClass.is(ModuleClass), moduleClass)
-      report.debuglog(s"Dumping mirror class for object: $moduleClass")
 
       val linkedClass  = moduleClass.companionClass
-      lazy val conflictingNames: Set[Name] = {
+      lazy val conflictingNames: Set[Name] =
         (linkedClass.info.allMembers.collect { case d if d.name.isTermName => d.name }).toSet
-      }
-      report.debuglog(s"Potentially conflicting names for forwarders: $conflictingNames")
-
-      for (m0 <- sortedMembersBasedOnFlags(moduleClass.info, required = Method, excluded = ExcludedForwarder)) {
-        val m = if (m0.is(Bridge)) m0.nextOverriddenSymbol else m0
-        if (m == NoSymbol)
-          report.log(s"$m0 is a bridge method that overrides nothing, something went wrong in a previous phase.")
-        else if (m.isType || m.is(Deferred) || (m.owner eq defn.ObjectClass) || m.isConstructor || m.name.is(ExpandedName))
-          report.debuglog(s"No forwarder for '$m' from $jclassName to '$moduleClass'")
-        else if (conflictingNames(m.name))
-          report.log(s"No forwarder for $m due to conflict with ${linkedClass.info.member(m.name)}")
-        else if (m.accessBoundary(defn.RootClass) ne defn.RootClass)
-          report.log(s"No forwarder for non-public member $m")
-        else {
-          report.log(s"Adding static forwarder for '$m' from $jclassName to '$moduleClass'")
-          addForwarder(jclass, moduleClass, m)
-        }
-      }
+
+      // There's a bit of a dance to fix lampepfl/dotty#12753 in a binary compatible way (see lampepfl/dotty#12767).
+      // The binary-incompatible solution would be to look up members at erasure phase, but we want to continue
+      // emitting the excess forwarders and mark them `synthetic`.
+      // Because additional members are generated after erasure (mixin) it's difficult to compare lists of symbols
+      // at different phases and know which ones to mark synthetic.
```

Comment:
Given:
```scala
class A[T] { def get: T = ??? }
object T extends A[String] { override def get: String = "hi" }
```

If I look at what we do before this PR, I see that sortedMembersBasedOnFlags returns two `get` methods, but only one of them `.is(Bridge)`, can't we rely on that to know that its forwarder should be synthetic?

Reply:
I gave that a try, but

```scala
trait T { def m  = 1 }
object O extends T
```

looking up `O$`'s members at the backend phase we get a method `O$.d` that's marked `Bridge`. Also, for

```scala
trait T { def m: Object }
object O extends T { def m: String = "" }
```

the lookup returns both a bridge `()Object` and the `()String` method. However, 3.0.0 doesn't generate a static accessor for the bridge in this case, so it would be quite weird to start adding a new synthetic method now.

Reply:
> looking up O$'s members at the backend phase we get a method O$.d that's marked Bridge. 

Ah, that's because we mark mixin forwarders as bridge too, so we could try excluding only bridges generated at erasure.

> However, 3.0.0 doesn't generate a static accessor for the bridge in this case, so it would be quite weird to start adding a new synthetic method now.

But if we only add this check in the place where the original code called `addForwarder` we shouldn't end up adding more forwarders.

I've implemented this idea in https://github.com/lampepfl/dotty/pull/12860, let me know what you think.